### PR TITLE
JDK-8323543: NPE when table items are set to null

### PR DESCRIPTION
**Pull request by EstelonAgarwaen**
This PR fixes a nullpointer in TableSkinUtils that occured when the Tables items were null.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed (2 reviews required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer), 1 [Author](https://openjdk.org/bylaws#author))

### Issue
 * [JDK-8323543](https://bugs.openjdk.org/browse/JDK-8323543): NPE when table items are set to null (**Bug** - P4)

### Reviewers
 * [Andy Goryachev](https://openjdk.org/census#angorya) (@andy-goryachev-oracle - **Reviewer**)
 * [Marius Hanl](https://openjdk.org/census#mhanl) (@Maran23 - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jfx.git pull/1329/head:pull/1329` \
`$ git checkout pull/1329`

Update a local copy of the PR: \
`$ git checkout pull/1329` \
`$ git pull https://git.openjdk.org/jfx.git pull/1329/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1329`

View PR using the GUI difftool: \
`$ git pr show -t 1329`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jfx/pull/1329.diff">https://git.openjdk.org/jfx/pull/1329.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jfx/pull/1329#issuecomment-1885444337)

### CODE REVIEW

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java [hunk 1]
```diff
@@ -7250,4 +7253,26 @@ public void testFirstRowSelectionWithEmptyArrayAsParameter() {
...
+        for (int i = 0; i < 4; i++) {
+            TreeItem<String> parent = new TreeItem<String>("item - " + i);
+            treeTableView.getRoot().getChildren().add(parent);
+        }
+
+        stageLoader = new StageLoader(treeTableView);
+        treeTableView.setRoot(null);
+
+        assertDoesNotThrow(() -> Toolkit.getToolkit().firePulse());
+
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TreeTableViewTest.java:7272]: please remove extra newline

**Comment by andy-goryachev-oracle** [comment]
FYI: there is also a possibility of an NPE in 

TableCellBehavior:70
TableView:2198
TableView:1922

I think we should add null checks there as well.

**Comment by EstelonAgarwaen** [comment]
should i add the null checks in this PR?

**Comment by andy-goryachev-oracle** [comment]
I think it makes sense to add the checks in this PR.  We are addressing the issue and not introducing new code paths.
I wonder if we should also add the unit tests to cover the newly added checks.  What do you think?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java [hunk 2]
```diff
@@ -6064,4 +6065,19 @@ public void testFirstRowSelectionWithEmptyArrayAsParameter() {
...
+        c.setCellValueFactory(value -> new SimpleStringProperty(value.getValue()));
+        table.getColumns().add(c);
+
+        table.getItems().addAll("1", "2", "3");
+
+        stageLoader = new StageLoader(table);
+        table.setItems(null);
+
+        assertDoesNotThrow(() -> Toolkit.getToolkit().firePulse());
+
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java:6081]: extra newline?

**Comment by EstelonAgarwaen** [comment]
i will look into whether it is possible to generate a nullpointer with the current code, and add a test to check that they are not thrown anymore.

**Comment by andy-goryachev-oracle** [comment]
Thank you @EstelonAgarwaen and @Maran23

**Comment by EstelonAgarwaen** [comment]
Regarding the NPEs: 
- i was able to trigger an exception in TableView:1922
- the code in TableCellBehaviour:70 is only reachable by user interaction with a cell. A cell only exists when a TableView has items. The interaction will never happen while `items` is null, because the code is unreachable in that case

**Comment by andy-goryachev-oracle** [comment]
I might suggest to add a null check to TableCellBehaviour:70 regardless, just to be on the safe side.

**Comment by EstelonAgarwaen** [comment]
since the conditions needed for TableCellBehaiviour:70 to be execute require items to not be null for that code to even be triggered, i think it does not make sense to check whether the items are null.

**Review by andy-goryachev-oracle — APPROVED** [review]
Looks good, thank you for finding and fixing the issue!

This PR has a very narrow focus to be a candidate for a back port to jfx22 @kevinrushforth

**Review by andy-goryachev-oracle — APPROVED** [review]
FYI, pushing another commit requires re-review (easy in this particular case).

**Comment by EstelonAgarwaen** [comment]
dont worry, im in a call with @Maran23 rn

**Comment by andy-goryachev-oracle** [comment]
@Maran23 would you like to double check and sponsor?

**Comment by Maran23** [comment]
> @Maran23 would you like to double check and sponsor?

sure!

**Comment by kevinrushforth** [comment]
> This PR has a very narrow focus to be a candidate for a back port to jfx22 @kevinrushforth

This seems a safe and targeted fix. I have no objection.

**Comment by kevinrushforth** [comment]
Although... I just spotted something. The updated tests use JUnit5 assertions into what is otherwise a JUnit4 test. We generally avoid doing this.

**Comment by Maran23** [comment]
> Although... I just spotted something. The updated tests use JUnit5 assertions into what is otherwise a JUnit4 test. We generally avoid doing this.

Should we use the JUnit5 `@Test` annotation, or migrate the class to JUnit5? The migration is also fine, although some assert need to be switched from `assertEquals(text, expected, actual)` -> `assertEquals(expected, actual, text)`

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java [hunk 3]
```diff
@@ -3871,22 +3929,22 @@ private void test_rt_37538(boolean callCNextOnce, boolean callCNextInLoop) {
         sl.dispose();
     }
 
-    @Ignore("Fix not yet developed for TableView")
+    @Disabled("Fix not yet developed for TableView")
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java:3932]: `@Test` was placed on its own line elsewhere but not in this file?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java [hunk 4]
```diff
@@ -5334,8 +5411,10 @@ public void test_rt_40012_selectedItemNotificationOnDisjointRemovesAbove() {
 
     @Test public void test_rt_40319_toRight_toBottom()          { test_rt_40319(true, true, false);   }
     @Test public void test_rt_40319_toRight_toTop()             { test_rt_40319(true, false, false);  }
-    @Test public void test_rt_40319_toLeft_toBottom()           { test_rt_40319(false, true, false);  }
-    @Test public void test_rt_40319_toLeft_toTop()              { test_rt_40319(false, false, false); }
+    @Test
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java:5414]: inconsistent formatting?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java [hunk 5]
```diff
@@ -5695,7 +5780,8 @@ public void test_clearAndSelectChangeMultipleSelectionCellMode() {
     @Test
     public void test_ChangeToStringKeyboardMultipleSelectionCellMode() {
         final Thread.UncaughtExceptionHandler exceptionHandler = Thread.currentThread().getUncaughtExceptionHandler();
-        Thread.currentThread().setUncaughtExceptionHandler((t, e) -> fail("We don't expect any exceptions in this test!"));
+        Thread.currentThread().setUncaughtExceptionHandler((t, e) -> fail(
+                "We don't expect any exceptions in this test!"));
```
Comment (andy-goryachev-oracle) [inline modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java:5784]: unrelated change?

File: modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java (hunk 3 above)
Comment (Maran23) [reply modules/javafx.controls/src/test/java/test/javafx/scene/control/TableViewTest.java:3932]: Hmm, I suggested @EstelonAgarwaen to use the IntelliJ Junit 5 Migration functionality. So looks like it is not always putting `@Test` in a new line. Only when nothing is above it seems?

**Comment by andy-goryachev-oracle** [comment]
I wonder if it was a mistake to migrate to junit5 in this PR, it should have been a separate task: the scope of the change has exploded, it might cause issues with backporting, etc.

Would it be possible to separate the migration from the fix?

**Comment by Maran23** [comment]
> Would it be possible to separate the migration from the fix?

Another idea could be to use the full qualified path: `@org.junit.jupiter.api.Test`. But `@Before` and `@After` are still JUnit 4.
Or we can reduce the change by at least reverting the new line changes? Then the diff should be much smaller.

**Comment by andy-goryachev-oracle** [comment]
I think the problem is that we are trying to address two issues in one PR - the fix and migration to junit5.
It might be better to go back one commit (but keep the most recent changes on a separate branch), convert the fix to use junit4 and submit the fix in this PR.

Then, create a separate JBS ticket for junit5 migration.  This way the scope of the fix is reduced to a minimum, and it can be easily backported and not cause many issues.

**Comment by Maran23** [comment]
> I think the problem is that we are trying to address two issues in one PR - the fix and migration to junit5. It might be better to go back one commit (but keep the most recent changes on a separate branch), convert the fix to use junit4 and submit the fix in this PR.

Sure, just FYI that there is nothing like `assertDoesNotThrow` for JUnit4.
What I did in the past is to at least write a comment, so we can easily add/convert that later on.
See also my PR in the past:
https://github.com/openjdk/jfx/pull/557
More specific:
https://github.com/openjdk/jfx/pull/557/files#diff-ea6f98c080f3abde83c8de64cd1b026b727bd5998cf6e0e57dde857f0e3e541eR353
cc @EstelonAgarwaen, I would recommend doing it the same way I did above and stash the JUnit 5 migration, as @andy-goryachev-oracle mentioned.

**Comment by Maran23** [comment]
> looks good now, thanks! should the fix be backported to jfx22?

Sounds good. I can do that, just need to find the instructions again.

**Comment by andy-goryachev-oracle** [comment]
backport instructions:
https://openjdk.org/guide/#working-with-backports-in-jbs

@kevinrushforth there is still time to backport, right?

**Comment by kevinrushforth** [comment]
> backport instructions: https://openjdk.org/guide/#working-with-backports-in-jbs

See also https://mail.openjdk.org/pipermail/openjfx-dev/2024-January/044529.html which has additional information.

> @kevinrushforth there is still time to backport, right?

Yes. JavaFX 22 is in RDP1 for another 3 weeks.

**Comment by kevinrushforth** [comment]
> /backport jfx22u

@Maran23 That will target jfx22u for JavaFX 22.0.1, which would probably be OK, but as I mentioned above, since we are still in RDP1 for JavaFX 22, this is a good candidate for JavaFX 22. So I recommend abandoning the above PR and instead target the jfx22 branch of the jfx repo: `/backport jfx jfx22` instead.

**Comment by Maran23** [comment]
> So I recommend abandoning the above PR and instead target the jfx22 branch of the jfx repo: `/backport jfx jfx22` instead.

Ahh, makes sense, thanks. First time backporting. :)
I deleted the backport branch in jfx22u.

**Comment by kevinrushforth** [comment]
> Ahh, makes sense, thanks. First time backporting. :)

Yeah, and it's somewhat confusing to have jfx22u active during RDP1 of JavaFX 22 where most fixes that go back into the 22 code line will want to go into 22 GA rather than 22.0.1. I had considered delaying the announcement of jfx22u being open, but we need it to backport the WebKit update (which we explicitly do not want in JavaFX 22 GA, but in 22.0.1).

**Comment by Maran23** [comment]
> Yeah, and it's somewhat confusing to have jfx22u ..., but we need it to backport the WebKit update (which we explicitly do not want in JavaFX 22 GA, but in 22.0.1).

Thanks for the clarification!
I totally forgot/missed that we have a branch for every GA release in this repository, while every backport after is done in the jfx..u repository.
